File: library/src/main/java/com/afollestad/aesthetic/views/AestheticCoordinatorLayout.java
package com.afollestad.aesthetic.views;

import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.view.menu.ActionMenuItemView;
import android.support.v7.widget.ActionMenuView;
import android.support.v7.widget.Toolbar;
import android.util.AttributeSet;
import android.view.Menu;
import android.view.View;

import com.afollestad.aesthetic.Aesthetic;
import com.afollestad.aesthetic.TintHelper;
import com.afollestad.aesthetic.Util;

import java.lang.reflect.Field;

import rx.Subscription;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;
import static com.afollestad.aesthetic.Rx.distinctToMainThread;
import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;

/** @author Aidan Follestad (afollestad) */
@RestrictTo(LIBRARY_GROUP)
public class AestheticCoordinatorLayout extends CoordinatorLayout
    implements AppBarLayout.OnOffsetChangedListener {

  private Subscription statusBarColorSubscription;
  private AppBarLayout appBarLayout;
  private View colorView;
  private AestheticToolbar toolbar;
  private CollapsingToolbarLayout collapsingToolbarLayout;
  private int toolbarColor;
  private int lastOffset = -1;

  public AestheticCoordinatorLayout(Context context) {
    super(context);
  }

  public AestheticCoordinatorLayout(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AestheticCoordinatorLayout(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  @SuppressWarnings("unchecked")
  private static void tintMenu(
      @NonNull AestheticToolbar toolbar, @Nullable Menu menu, final @ColorInt int color) {
    if (toolbar.getNavigationIcon() != null) {
      toolbar.setNavigationIcon(toolbar.getNavigationIcon(), color);
    }
    Util.setOverflowButtonColor(toolbar, color);

    try {
      final Field field = Toolbar.class.getDeclaredField("mCollapseIcon");
      field.setAccessible(true);
      Drawable collapseIcon = (Drawable) field.get(toolbar);
      if (collapseIcon != null) {
        field.set(toolbar, TintHelper.createTintedDrawable(collapseIcon, color));
      }
    } catch (Exception e) {
      e.printStackTrace();
    }

    final PorterDuffColorFilter colorFilter =
        new PorterDuffColorFilter(color, PorterDuff.Mode.SRC_IN);
    for (int i = 0; i < toolbar.getChildCount(); i++) {
      final View v = toolbar.getChildAt(i);
      // We can't iterate through the toolbar.getMenu() here, because we need the ActionMenuItemView.
      if (v instanceof ActionMenuView) {
        for (int j = 0; j < ((ActionMenuView) v).getChildCount(); j++) {
          final View innerView = ((ActionMenuView) v).getChildAt(j);
          if (innerView instanceof ActionMenuItemView) {
            int drawablesCount = ((ActionMenuItemView) innerView).getCompoundDrawables().length;
            for (int k = 0; k < drawablesCount; k++) {
              if (((ActionMenuItemView) innerView).getCompoundDrawables()[k] != null) {
                ((ActionMenuItemView) innerView)
                    .getCompoundDrawables()[k].setColorFilter(colorFilter);
              }
            }
          }
        }
      }
    }

    if (menu == null) {
      menu = toolbar.getMenu();
    }
    ViewUtil.tintToolbarMenu(toolbar, menu, color);
  }

  @Override
  public void onAttachedToWindow() {
    super.onAttachedToWindow();

    // Find the toolbar and color view used to blend the scroll transition
    if (getChildCount() > 0 && getChildAt(0) instanceof AppBarLayout) {
      appBarLayout = (AppBarLayout) getChildAt(0);
      if (appBarLayout.getChildCount() > 0
          && appBarLayout.getChildAt(0) instanceof CollapsingToolbarLayout) {
        collapsingToolbarLayout = (CollapsingToolbarLayout) appBarLayout.getChildAt(0);
        for (int i = 0; i < collapsingToolbarLayout.getChildCount(); i++) {
          if (this.toolbar != null && this.colorView != null) {
            break;
          }
          View child = collapsingToolbarLayout.getChildAt(i);
          if (child instanceof AestheticToolbar) {
            this.toolbar = (AestheticToolbar) child;
          } else if (child.getBackground() != null
              && child.getBackground() instanceof ColorDrawable) {
            this.colorView = child;
          }
        }
      }
    }

    if (toolbar != null && colorView != null) {
      this.appBarLayout.addOnOffsetChangedListener(this);
      toolbar
          .colorUpdated()
          .compose(distinctToMainThread())
          .subscribe(
              color -> {
                toolbarColor = color;
                invalidateColors();
              },
              onErrorLogAndRethrow());
    }

    statusBarColorSubscription =
        Aesthetic.get()
            .statusBarColor()
            .compose(distinctToMainThread())
            .subscribe(
                color -> {
                  collapsingToolbarLayout.setContentScrimColor(color);
                  collapsingToolbarLayout.setStatusBarScrimColor(color);
                },
                onErrorLogAndRethrow());
  }

  @Override
  public void onDetachedFromWindow() {
    statusBarColorSubscription.unsubscribe();
    this.appBarLayout.removeOnOffsetChangedListener(this);
    this.appBarLayout = null;
    this.toolbar = null;
    this.colorView = null;
    super.onDetachedFromWindow();
  }

  @Override
  public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
    if (lastOffset == Math.abs(verticalOffset)) {
      return;
    }
    lastOffset = Math.abs(verticalOffset);
    invalidateColors();
  }

  private void invalidateColors() {
    final int maxOffset = appBarLayout.getMeasuredHeight() - toolbar.getMeasuredHeight();
    final float ratio = (float) lastOffset / (float) maxOffset;

    final int colorViewColor = ((ColorDrawable) colorView.getBackground()).getColor();
    final int blendedColor = Util.blendColors(colorViewColor, toolbarColor, ratio);
    final int collapsedTitleColor = Util.isColorLight(toolbarColor) ? Color.BLACK : Color.WHITE;
    final int expandedTitleColor = Util.isColorLight(colorViewColor) ? Color.BLACK : Color.WHITE;
    final int blendedTitleColor = Util.blendColors(expandedTitleColor, collapsedTitleColor, ratio);

    toolbar.setBackgroundColor(blendedColor);

    collapsingToolbarLayout.setCollapsedTitleTextColor(collapsedTitleColor);
    collapsingToolbarLayout.setExpandedTitleColor(expandedTitleColor);

    tintMenu(toolbar, toolbar.getMenu(), blendedTitleColor);
  }
}

File: library/src/main/java/com/afollestad/aesthetic/views/AestheticCoordinatorLayout.kt
/**
 * Designed and developed by Aidan Follestad (@afollestad)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.afollestad.aesthetic.views

import android.content.Context
import android.graphics.Color
import android.graphics.Color.BLACK
import android.graphics.Color.WHITE
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.Menu
import android.view.View
import androidx.appcompat.view.menu.ActionMenuItemView
import androidx.appcompat.widget.ActionMenuView
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout
import com.afollestad.aesthetic.ActiveInactiveColors
import com.afollestad.aesthetic.Aesthetic.Companion.get
import com.afollestad.aesthetic.utils.adjustAlpha
import com.afollestad.aesthetic.utils.blendWith
import com.afollestad.aesthetic.utils.distinctToMainThread
import com.afollestad.aesthetic.utils.findField
import com.afollestad.aesthetic.utils.isColorLight
import com.afollestad.aesthetic.utils.setOverflowButtonColor
import com.afollestad.aesthetic.utils.subscribeTo
import com.afollestad.aesthetic.utils.tint
import com.afollestad.aesthetic.utils.tintMenu
import com.afollestad.aesthetic.utils.unsubscribeOnDetach
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout

/** @author Aidan Follestad (afollestad) */
class AestheticCoordinatorLayout(
  context: Context,
  attrs: AttributeSet? = null
) : CoordinatorLayout(context, attrs), AppBarLayout.OnOffsetChangedListener {

  companion object {
    const val UNFOCUSED_ALPHA = 0.5f
  }

  private var appBarLayout: AppBarLayout? = null
  private var colorView: View? = null
  private var toolbar: AestheticToolbar? = null
  private var collapsingToolbarLayout: CollapsingToolbarLayout? = null

  private var toolbarColor: Int = 0
  private var iconTextColors: ActiveInactiveColors? = null
  private var lastOffset = -1

  private fun tintMenu(
    toolbar: AestheticToolbar,
    menu: Menu?,
    colors: ActiveInactiveColors
  ) {
    if (toolbar.navigationIcon != null) {
      toolbar.setNavigationIcon(toolbar.navigationIcon, colors.activeColor)
    }
    toolbar.setOverflowButtonColor(colors.activeColor)

    try {
      val field = Toolbar::class.findField("mCollapseIcon")
      val collapseIcon = field.get(toolbar) as? Drawable
      if (collapseIcon != null) {
        field.set(toolbar, collapseIcon.tint(colors.toEnabledSl()))
      }
    } catch (e: Exception) {
      e.printStackTrace()
    }

    val colorFilter = PorterDuffColorFilter(colors.activeColor, PorterDuff.Mode.SRC_IN)
    for (i in 0 until toolbar.childCount) {
      val v = toolbar.getChildAt(i)
      // We can't iterate through the toolbar.getMenu() here, because we need the
      // ActionMenuItemView.
      if (v is ActionMenuView) {
        for (j in 0 until v.childCount) {
          val innerView = v.getChildAt(j)
          if (innerView is ActionMenuItemView) {
            val drawablesCount = innerView.compoundDrawables.size
            for (k in 0 until drawablesCount) {
              if (innerView.compoundDrawables[k] != null) {
                innerView
                    .compoundDrawables[k].colorFilter = colorFilter
              }
            }
          }
        }
      }
    }
    toolbar.tintMenu(menu ?: toolbar.menu, colors.activeColor, colors.inactiveColor)
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()

    // Find the toolbar and color view used to blend the scroll transition
    if (childCount > 0 && getChildAt(0) is AppBarLayout) {
      appBarLayout = getChildAt(0) as AppBarLayout
      if (appBarLayout!!.childCount > 0 && appBarLayout!!.getChildAt(
              0
          ) is CollapsingToolbarLayout
      ) {
        collapsingToolbarLayout = appBarLayout!!.getChildAt(0) as CollapsingToolbarLayout
        for (i in 0 until collapsingToolbarLayout!!.childCount) {
          if (this.toolbar != null && this.colorView != null) {
            break
          }
          val child = collapsingToolbarLayout!!.getChildAt(i)
          if (child is AestheticToolbar) {
            this.toolbar = child
          } else if (child.background != null && child.background is ColorDrawable) {
            this.colorView = child
          }
        }
      }
    }

    if (toolbar != null && colorView != null) {
      this.appBarLayout?.addOnOffsetChangedListener(this)

      toolbar!!.colorUpdated()
          .distinctToMainThread()
          .subscribeTo {
            toolbarColor = it
            val baseColor = if (it.isColorLight()) BLACK else WHITE
            iconTextColors = ActiveInactiveColors(baseColor, baseColor.adjustAlpha(UNFOCUSED_ALPHA))
            invalidateColors()
          }
          .unsubscribeOnDetach(this)
    }

    if (collapsingToolbarLayout != null) {
      get().colorStatusBar()
          .distinctToMainThread()
          .subscribeTo {
            collapsingToolbarLayout?.apply {
              setContentScrimColor(it)
              setStatusBarScrimColor(it)
            }
          }
          .unsubscribeOnDetach(this)
    }
  }

  override fun onDetachedFromWindow() {
    this.appBarLayout?.removeOnOffsetChangedListener(this)
    this.appBarLayout = null
    this.toolbar = null
    this.colorView = null
    super.onDetachedFromWindow()
  }

  override fun onOffsetChanged(
    appBarLayout: AppBarLayout,
    verticalOffset: Int
  ) {
    if (lastOffset == Math.abs(verticalOffset)) return
    lastOffset = Math.abs(verticalOffset)
    invalidateColors()
  }

  private fun invalidateColors() {
    if (iconTextColors == null) {
      return
    }

    val maxOffset = appBarLayout!!.measuredHeight - toolbar!!.measuredHeight
    val ratio = lastOffset.toFloat() / maxOffset.toFloat()

    val colorViewColor = (colorView!!.background as ColorDrawable).color
    val blendedColor = colorViewColor.blendWith(toolbarColor, ratio)
    val collapsedTitleColor = iconTextColors!!.activeColor
    val expandedTitleColor = if (colorViewColor.isColorLight()) Color.BLACK else Color.WHITE
    val blendedTitleColor = expandedTitleColor.blendWith(collapsedTitleColor, ratio)

    toolbar?.apply {
      setBackgroundColor(blendedColor)
      tintMenu(
          this,
          menu,
          ActiveInactiveColors(
              blendedTitleColor,
              blendedColor.adjustAlpha(0.7f)
          )
      )
    }

    collapsingToolbarLayout?.apply {
      setCollapsedTitleTextColor(collapsedTitleColor)
      setExpandedTitleColor(expandedTitleColor)
    }
  }
}
